### PR TITLE
esp8266: Some API additions + misc fixes

### PR DESCRIPTION
**Pull request by atx**

### CODE REVIEW

**Comment by pfalcon** [comment]
> esp8266: Enable setting CPU frequency to 160MHz	

Did you test stability of this? In standard SDK, such frequency is enabled only for short periods of time for hard number crunching (SSL).

**Comment by atalax** [comment]
> Did you test stability of this?

I have not done any extensive testing, but the chip does not seem to be heating (more than usual) or crashing. There are no mentions of problems on the internet too, so it is probably safe.

**Comment by dpgeorge** [comment]
Thanks @atalax, lots of new features!  But parts of this PR will need some thought/discussion first.

Mainly, there is existing API for things like ADC, unique id, version number, etc, and we should try to use these "standards" instead of inventing new ones.

File: esp8266/modesp.c [hunk 1]
```diff
@@ -37,535 +37,57 @@
...
-STATIC mp_obj_t esp_socket___del__(mp_obj_t self_in) {
-    esp_socket_obj_t *s = self_in;
-
-    esp_socket_close(self_in);
-
-    if (s->fromserver) {
-        espconn_delete(s->espconn);
-    }
-
+STATIC mp_obj_t esp_deepsleep(mp_uint_t n_args, const mp_obj_t *args) {
```
Comment (dpgeorge) [inline esp8266/modesp.c]: We don't yet have a standard for sleep modes (pyboard has pyb.stop(), cc3200 has pyb.Sleep() class, and there is a discussion to put it in the machine module), so this can stay here for now.

File: esp8266/modesp.c [hunk 2]
```diff
@@ -37,535 +37,57 @@
...
+STATIC MP_DEFINE_CONST_FUN_OBJ_VAR_BETWEEN(esp_deepsleep_obj, 0, 1, esp_deepsleep);
 
-STATIC void esp_socket_sent_callback(void *arg) {
-    struct espconn *conn = arg;
-    esp_socket_obj_t *s = conn->reverse;
-
-    if (s->cb_sent != mp_const_none) {
-        call_function_1_protected(s->cb_sent, s);
-    }
+STATIC mp_obj_t esp_adc() {
```
Comment (dpgeorge) [inline esp8266/modesp.c]: ADC capability is provided by pyb.ADC() class.  Yes that's overkill for ESP which seems to have only 1 ADC pin, but would be best to provide some consistency here, no?

File: esp8266/modesp.c [hunk 3]
```diff
@@ -37,535 +37,57 @@
...
-STATIC void esp_socket_disconnect_callback(void *arg) {
-    struct espconn *conn = arg;
-    esp_socket_obj_t *s = conn->reverse;
-
-    if (s->cb_disconnect != mp_const_none) {
-        call_function_1_protected(s->cb_disconnect, s);
-    }
-
-    esp_socket_close(s);
+STATIC mp_obj_t esp_vdd33() {
```
Comment (dpgeorge) [inline esp8266/modesp.c]: This could be considered as ADC read of an internal channel.

File: esp8266/modesp.c [hunk 4]
```diff
@@ -37,535 +37,57 @@
...
-    espconn_regist_sentcb(conn, esp_socket_sent_callback);
-    espconn_regist_disconcb(conn, esp_socket_disconnect_callback);
-    espconn_regist_time(conn, 15, 0);
-
-    if (esp_socket_listening->cb_connect != mp_const_none) {
-        call_function_1_protected(esp_socket_listening->cb_connect, s);
-    } else {
-        mp_obj_list_append(esp_socket_listening->connlist, s);
-    }
+STATIC mp_obj_t esp_chip_id() {
```
Comment (dpgeorge) [inline esp8266/modesp.c]: pyboard has pyb.unique_id() for this.

File: esp8266/modesp.c [hunk 5]
```diff
@@ -37,535 +37,57 @@
...
-            "not connected"));
-    }
-
-    mp_buffer_info_t bufinfo;
-    mp_get_buffer_raise(buf_in, &bufinfo, MP_BUFFER_READ);
-
-    espconn_sent(s->espconn, bufinfo.buf, bufinfo.len);
-
-    return mp_obj_new_int(bufinfo.len);
+STATIC mp_obj_t esp_sdk_version() {
```
Comment (dpgeorge) [inline esp8266/modesp.c]: This might fit better in os.uname().

File: esp8266/modespwifi.c [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+
+STATIC mp_obj_t esp_wifi_scan(mp_obj_t cb_in) {
+    MP_STATE_PORT(scan_cb_obj) = cb_in;
+    wifi_set_opmode(STATION_MODE);
+    wifi_station_scan(NULL, (scan_done_cb_t)esp_wifi_scan_cb);
+    return mp_const_none;
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_1(esp_wifi_scan_obj, esp_wifi_scan);
+
+STATIC mp_obj_t esp_wifi_ip(mp_uint_t n_args, const mp_obj_t *args) {
```
Comment (dpgeorge) [inline esp8266/modespwifi.c]: cc3200 has `ifconfig` method which returns a (named) tuple of results, including IP and MAC addresses.  It also has `config_ip` to set the IP etc.

@danicampora would you consider merging config_ip into ifconfig, such that if no args are given to ifconfig then it queries the config, otherwise it sets it?  This would save you a few bytes.

File: esp8266/qstrdefsport.h [hunk 7]
```diff
@@ -42,11 +42,23 @@ Q(sync)
 Q(hard_reset)
 
 Q(esp)
+Q(esp.wifi)
```
Comment (dpgeorge) [inline esp8266/qstrdefsport.h]: Does this work, having a dot in the name?  It's certainly not intended to work!  @pfalcon can we allow this and make it a "feature" of the way import works?

File: esp8266/modespwifi.c [hunk 8]
```diff
@@ -0,0 +1,223 @@
...
+    if (n_args == 0) {
+        return mp_obj_new_int(wifi_get_sleep_type());
+    } else {
+        wifi_set_sleep_type(mp_obj_get_int(args[0]));
+        return mp_const_none;
+    }
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_VAR_BETWEEN(esp_wifi_sleep_type_obj, 0, 1, esp_wifi_sleep_type);
+
+STATIC mp_obj_t esp_wifi_mac(mp_uint_t n_args, const mp_obj_t *args) {
```
Comment (dpgeorge) [inline esp8266/modespwifi.c]: cc3200 just returns raw bytes object of the MAC address.  That saves a lot of bytes not having to make a function to parse/format the address (which is not as common as formatting/parsing an IP address).  And if you really need to format it then you can write one line of Python to do it.

File: esp8266/qstrdefsport.h (hunk 7 above)
Comment (pfalcon) [inline esp8266/qstrdefsport.h]: If it works, it's kinda bug, and likely against Python semantics. If it's useful, worth thinking.

P.S. Actually, the only way it might work is if parser would parse it as one qstr. Does it? Again, that would be a bug in parser ;-).

File: esp8266/modespwifi.c (hunk 6 above)
Comment (danicampora) [inline esp8266/modespwifi.c]: > @danicampora would you consider merging config_ip into ifconfig, such that if no args are given to ifconfig then it queries the config, otherwise it sets it? This would save you a few bytes.

Yes of course, if you think that makes more sense to use ifconfig for both things, then I'm happy with it :-).

Comment (dpgeorge) [inline esp8266/modespwifi.c]: Well, you have as much say as I do here :)  But actually I think it's a good idea, and in fact that's how the pyboard's wiznet driver works.

Comment (danicampora) [inline esp8266/modespwifi.c]: If I merge my two functions into ifconfig, the API would be:
`ifconfig(mode, *, ip='192.168.1.1', subnet='255.255.255.0', gateway='192.168.1.1', dns='8.8.8.8')` to set
and `ifconfig()` to get it.

However, the wiznet driver uses:
`ifconfig([(ip, subnet, gateway, dns)])` which I like, but doesn't allow to set the IP mode to dynamic. I propose the following for the CC3200/ESP8266:

`ifconfig((ip, subnet, gateway, dns))` sets a static IP.
`ifconfig(())` with an empty tuple sets the IP mode to dynamic.
`ifconfig()` gets the interface config

Thoughts?

Comment (dpgeorge) [inline esp8266/modespwifi.c]: Move to issue #1269.

**Comment by atalax** [comment]
Updated, I also removed `ip()` until #1269 gets implemented in cc3200

**Comment by atalax** [comment]
Ok, everything should be in order now

**Comment by roger-** [comment]
Any idea when this'll get merged?

**Comment by pfalcon** [comment]
Merging big changes is always hard. Merging changes which involve splits and rename is well, pretty hard - there should be reason provided why that's needed, which is not there.

@atalax : I cherry-picked "esp8266: Add uos module", please rebase.

**Comment by pfalcon** [comment]
Merged "esp8266: Move initialization to system_init_done_cb".

@atalax : Please confirm that ADC stuff is tested.

**Comment by pfalcon** [comment]
@dpgeorge : Let's discuss https://github.com/atalax/micropython/commit/5214cb0fde2207213982e7dd066c92ab634d0ea3 "esp8266: Enable setting CPU frequency to 160MHz". So, this gets rid of a HAL function and calls vendor API directly from a micropython module. I'm +1 on that, I don't see why we should maintain any "HAL", I assume for pyboard it's just means to organize hw-related code, not that we try to maintain some kind of "HAL" API. But maybe I'm missing something.

**Comment by atalax** [comment]
> - there should be reason provided why that's needed, which is not there.

I was kind off afraid that eventually, it would be hard to recognize which functions do what (for example, it is not exactly obvious that `esp.status()` and `esp.sleep_type()` are wifi-related)

>  Please confirm that ADC stuff is tested.

Works. Note that for reading vdd33:

```
The 107th byte in esp_init_data_default.bin（0〜127byte）is named as
“vdd33_const“ , when TOUT pin is suspended vdd33_const must be set as
0xFF， that is 255
```

This is probably supposed to be done by manufacturer of the board, but I have no idea to what extent it actually is.

> please rebase.

Done

**Comment by pfalcon** [comment]
>> The 107th byte in esp_init_data_default.bin（0〜127byte）is named as
“vdd33_const“ , when TOUT pin is suspended vdd33_const must be set as
0xFF， that is 255

> This is probably supposed to be done by manufacturer of the board, but I have no idea to what extent it actually is.

It's a bit hard to understand what the original quote is about, but to make sure we don't spur more confusion then it already has: why do you think anything there should be done by  manufacturer of the board?

**Comment by atalax** [comment]
> It's a bit hard to understand what the original quote is about, but to make sure we don't spur more confusion then it already has: why do you think anything there should be done by manufacturer of the board?

If I understand the programming guide correctly, it seems that measuring vdd and tout voltage is mutually exclusive. As it depends on the board whether tout is used, the manufacturer should set the byte accordingly on initial firmware load.

**Comment by pfalcon** [comment]
> As it depends on the board whether tout is used, the manufacturer should set the byte accordingly on initial firmware load.

Ok, to clarify, with esp8266, you're that manufacturer. What that excerpt refers to is esp_init_data_default.bin file in the SDK. Contents of that file are programmed somewhere closer to the end of the (external, SPI) flash, and any user can do it anytime (and sometimes it needs to be done, as esp8266 stores runtime params in that area too, and sometimes it can get into a weird state). As for hardware capabilities and actual semantics of this or another config byte, I can't comment, only experimentation and disassembly could show ;-).

**Comment by dpgeorge** [comment]
> Let's discuss atalax@5214cb0. So, this gets rid of a HAL function and calls vendor API directly from a micropython module. I'm +1 on that

Yes, I'm fine with it also.

> I don't see why we should maintain any "HAL"

For stdio stuff it's necessary so that, eg, readline and pyexec can be used across multiple ports.

> I assume for pyboard it's just means to organize hw-related code, not that we try to maintain some kind of "HAL" API. But maybe I'm missing something.

Yes, for things that are not cross-port it's a way of collecting board-related code.  Eg for pic16bit, mp_hal_milli_delay is just the name of the delay function.

For the specific function at hand, mp_hal_get_cpu_freq(), I did it that way in the ESP port because at that time you couldn't include the ESP headers in uPy code because of clashes with stdio (or something like that).  But since this function is only ever used by modpyb.c we don't need it and can call the system provided function directly.

Summary: this patch is good to merge as-is.

**Comment by pfalcon** [comment]
"esp8266: Add a bunch of miscellaneous methods" - I'm merging this, but as preliminary implementation, subject to change (e.g. SLEEP_MODEM is a bit weird from user friendliness perspective. Modem? What modem? Well, certainly there's modulator-demodulator somewhere there. But calling setting SLEEP_WIFI or SLEEP_RF makes more sense.)

**Comment by pfalcon** [comment]
Please rebase.

**Comment by atalax** [comment]
> Please rebase.

Done

**Comment by chinadsf** [comment]
Build with esp8266 SDK1.1 error:

    Use make V=1 or set BUILD_VERBOSE in your environment to increase build verbosity.
    LINK build/firmware.elf
    build/modpyb.o:(.text.pyb_freq+0x4): undefined reference to `system_get_cpu_freq'
    build/modpyb.o:(.text.pyb_freq+0x8): undefined reference to `system_update_cpu_freq'
    build/modpyb.o:(.text.pyb_freq+0x13): undefined reference to `system_get_cpu_freq'
    build/modpyb.o: In function `pyb_freq':
    modpyb.c:(.text.pyb_freq+0x52): undefined reference to `system_update_cpu_freq'
    build/modesp.o:(.text.esp_sleep_type+0x0): undefined reference to `wifi_get_sleep_type'
    build/modesp.o:(.text.esp_sleep_type+0xb): undefined reference to `wifi_get_sleep_type'
    build/modpybadc.o:(.text.pyb_adc_read+0x0): undefined reference to `system_get_vdd33'
    build/modpybadc.o: In function `pyb_adc_read':
    modpybadc.c:(.text.pyb_adc_read+0x16): undefined reference to `system_get_vdd33'
    build/moduos.o:(.text.os_uname+0x4): undefined reference to `system_get_sdk_version'
    build/moduos.o: In function `os_uname':
    moduos.c:(.text.os_uname+0x21): undefined reference to `system_get_sdk_version'
    make: *** [build/firmware.elf] Error 1

**Comment by atalax** [comment]
> Build with esp8266 SDK1.1 error:

Strange, are you using SDK from (esp-open-sdk)[https://github.com/pfalcon/esp-open-sdk]?

**Comment by pfalcon** [comment]
Ok, sow what's left here: 

> esp8266: Split modesp.c …
> Also separated the wifi functions to esp.wifi submodule

2 points here:

1) Reviewing renames/splits is hard. Every contributor should keep this in mind and understand that submitting a split is a good way to have your patch stuck in review queue for month(s). Perhaps we should add this to contributor guidelines. The right way to handle renames/splits is to raise an RFC and wait for maintainers' approval. Otherwise, avoid doing any renames and splits - if your changes need them, that will be pointed out.

2) No, it's not ok to put wifi-related functionality in "esp.wifi". 2 sub-reasons:

2.1) As I elaborated previously "esp" is intended to be analog of "socket" module, but with a callback API. It's not called "espsocket" because it's also a bucket to drop other esp8266-related stuff which doesn't fit anywhere else. Drop, as a temporary bucket. No need to create structures and superstructures there. There's no "socket.wifi", so can't be "esp.wifi" either.

2.2) So, how to do it right? Here's the doc: http://docs.micropython.org/en/latest/library/network.html . All interface-level configuration in uPy happens via "network" module (@dpgeorge, @danicampora: is that still true?). So, feel free to introduce "network" module for esp8266 which follows interface as describe in the docs above. Start with just a module skeleton, move functions one by one to ease review. If not, keeping all stuff flat in "esp" is ok for time being.

Summing up: -2 from me on "esp8266: Split modesp.c"

**Comment by danicampora** [comment]
> 2.2) So, how to do it right? Here's the doc: http://docs.micropython.org/en/latest/library/network.html . All interface-level configuration in uPy happens via "network" module (@dpgeorge, @danicampora: is that still true?). So, feel free to introduce "network" module for esp8266 which follows interface as describe in the docs above. Start with just a module skeleton, move functions one by one to ease review. If not, keeping all stuff flat in "esp" is ok for time being.

Yes, that's still true :-)

**Comment by pfalcon** [comment]
"esp8266: Do not call espconn_create in constructor of esp.socket" is applied manually to pre-rename file, I couldn't preserve authorship due to path conflict.

**Comment by pfalcon** [comment]
> Yes, that's still true :-)

Thanks.


So, all patches except one above were merged. Not closing right away to catch any feedback/objections.

**Comment by pfalcon** [comment]
Closing.
